feat: new connector sample app anaplan as API based example

prateek11rai opened the pull request:
### Changelog
<!-- Provide a clear and concise description of the changes in this PR in bullet points -->
- _to be added_

### Additional context (e.g. screenshots, logs, links)
<!-- Provide a clear additional context, dependencies & links in bullet points -->
<!-- links could be jira, slack, docs, etc. -->
- _to be added_


### Checklist
<!-- Mark [x] the appropriate option, helps the reviewer to verify the changes -->
- [ ] Additional tests added
- [ ] All CI checks passed
- [ ] Relevant documentation updated

<!-- for any cautionary notes, use https://github.com/orgs/community/discussions/16925 -->


<!-- for any questions, reachout to #pod-app-framework or connect@atlan.com -->

AtMrun commented on connectors/anaplan/app/workflows.py:
```diff
@@ -0,0 +1,225 @@
+        logger.info("Starting Anaplan metadata extraction workflow")
+        logger.info(f"Workflow config received: {workflow_config}")
+
+        logger.info("Starting workflow execution")
```
duplicate

prateek11rai replied:
Removed.

AtMrun commented on connectors/anaplan/app/workflows.py:
```diff
@@ -0,0 +1,225 @@
+            result = await workflow.execute_activity_method(
+                activities_instance.get_workflow_args,
+                args=[workflow_config],
+                retry_policy=RetryPolicy(maximum_attempts=3, backoff_coefficient=2),
```
use const

AtMrun commented on connectors/anaplan/app/workflows.py:
```diff
@@ -0,0 +1,225 @@
+            # STEP 3: Set metadata filter state
+            logger.info("Executing set_metadata_filter_state")
+            await workflow.execute_activity_method(
+                activities_instance.set_metadata_filter_state,
```
can you add more explanation what this is doing

prateek11rai commented on connectors/anaplan/app/workflows.py:
```diff
@@ -0,0 +1,225 @@
+        }
+        """
+        logger.info("Starting Anaplan metadata extraction workflow")
+        logger.info(f"Workflow config received: {workflow_config}")
```
Removed.

prateek11rai commented on connectors/anaplan/app/workflows.py:
```diff
@@ -0,0 +1,225 @@
+            workflow_args: Dict[str, Any] = {}  # Initialize workflow_args
+
+            # activity retry policy
+            retry_policy = RetryPolicy(maximum_attempts=2, backoff_coefficient=2)
```
Using :
```python
# Constants for activity retry policies
ACTIVITY_RETRY_MAX_ATTEMPTS = 3
ACTIVITY_RETRY_BACKOFF_COEFFICIENT = 2

...

retry_policy = RetryPolicy(maximum_attempts=ACTIVITY_RETRY_MAX_ATTEMPTS, backoff_coefficient=ACTIVITY_RETRY_BACKOFF_COEFFICIENT)
```

inishchith commented on connectors/anaplan/app/activities/__init__.py:
```diff
@@ -0,0 +1,318 @@
+            except json.JSONDecodeError:
+                raise ValueError("Invalid JSON in include-metadata or exclude-metadata")
+
+            return {
```
can you use pydantic models instead? we're moving away from using dictionaries. You can deserialize it in the workflow definition

prateek11rai replied:
Removed completely. Changed validations and removed decoder checks for now. Was making things complex.

ErnestoLoma replied:
@prateek11rai Did you commit your changes? I still see use of a dictionary?

inishchith commented on connectors/anaplan/app/handlers.py:
```diff
@@ -0,0 +1,570 @@
+    # SECTION 3: METADATA METHODS (Static helpers for fetch metadata)
+    # ============================================================================
+
+    async def _extract_apps_for_handler(self) -> List[Dict[str, Any]]:
```
shouldn't this be a reuse of the methods in `extracts/apps.py`?

inishchith commented on connectors/anaplan/app/handlers.py:
```diff
@@ -0,0 +1,570 @@
+            logger.error(f"Error extracting apps data for handler: {str(e)}")
+            raise
+
+    async def _extract_pages_for_handler(
```
same as above - shouldn't this be a reuse of the methods in `extracts/apps.py`?

inishchith commented on connectors/anaplan/pyproject.toml:
```diff
@@ -0,0 +1,114 @@
+]
+fail_under = 60
+
+[tool.uv.sources]
```
comment this please

prateek11rai commented on connectors/anaplan/app/activities/extracts/pages.py:
```diff
@@ -0,0 +1,266 @@
+            )
+
+            # Yield control back to event loop after each chunk
+            await asyncio.sleep(0)
```
removed.

prateek11rai commented on connectors/anaplan/app/clients.py:
```diff
@@ -0,0 +1,373 @@
+        self.host = credentials.get("host", "us1a.app.anaplan.com")
+        self.username = credentials.get("username")
+        self.password = credentials.get("password")
+        self.auth_type = credentials.get("authType", "basic")
```
removed.

prateek11rai commented:
Made recommended changes:
- Since sample apps have a dockerfile in root - did not add it in this specific sample app.
- re-wrote the doc strings in the recommended convention.
- Simplified the sample app with workflow run controlling the execution order of activities and get_activities used only for registering activities.
- Simplified by making app client use default retry transporter.
- Simplified by making extract functions re-usable.
- Added app-sdk recommended conventions for reading and writing files (object store).
- Made method names generic with removing the Anaplan prefix from them and calling them App.... . 
- Reduced the number of functions.
- Made validation simpler by only doing logical checks.

inishchith commented on connectors/anaplan/README.md:
```diff
@@ -0,0 +1,127 @@
+│   └── workflows/      # Workflow definitions and orchestration
+├── components/         # Dapr components (auto-downloaded)
+├── frontend/           # Web interface assets
+├── deploy/            # Installation and deployment files
```
please remove `deploy` we do not use it anymore

inishchith commented on connectors/anaplan/app/activities/__init__.py:
why not just name this `activities.py` inside /`app`?

ErnestoLoma commented on connectors/anaplan/app/activities/__init__.py:
```diff
@@ -0,0 +1,380 @@
+            app_data = await extract_apps_data(state.client)
+
+            # Check for filters and apply them
+            filtered_app_data = [
+                app
+                for app in app_data
+                if should_include_asset(
+                    app, "app", state.metadata_filter_state, state.metadata_filter
+                )
+            ]
+
+            # Create DataFrame and write to parquet
+            if filtered_app_data:
```
Cleaner to use a named expression as follows:

```
if filtered_app_data := [
                app
                for app in app_data
                if should_include_asset(
                    app, "app", state.metadata_filter_state, state.metadata_filter
                )
            ]:
```

vishnub-atlan replied:
@ErnestoLoma - nitpick, would refer readability given these are sample apps - not sure how knowledgeable people reading might be with python syntax ex `:=`, or a list comprehension.

ErnestoLoma commented on connectors/anaplan/app/activities/__init__.py:
```diff
@@ -0,0 +1,380 @@
+                    f"No apps found, skipping write: {parquet_output.get_full_path()}"
+                )
+
+            statistics = await parquet_output.get_statistics(typename="app")
```
No need for a variable, just return resutls

ErnestoLoma commented on connectors/anaplan/app/activities/__init__.py:
```diff
@@ -0,0 +1,380 @@
+            )
+
+            # Apply metadata filtering logic
+            detailed_page_data = [
```
Cleaner to use named expression:

```
if detailed_page_data := [
                page
                for page in all_detailed_page_data
                if should_include_asset(
                    page,
                    "page",
                    state.metadata_filter_state,
                    state.metadata_filter,
                )
]:
```

ErnestoLoma commented on connectors/anaplan/app/activities/__init__.py:
```diff
@@ -0,0 +1,380 @@
+                    f"No pages found, skipping write: {parquet_output.get_full_path()}"
+                )
+
+            statistics = await parquet_output.get_statistics(typename="page")
```
No need for a variable. Just return the results.

ErnestoLoma commented on connectors/anaplan/app/activities/__init__.py:
```diff
@@ -0,0 +1,380 @@
+                    raise ValueError(
+                        "Invalid JSON in include-metadata or exclude-metadata"
+                    )
+            except json.JSONDecodeError:
+                raise ValueError("Invalid JSON in include-metadata or exclude-metadata")
```
I think you should raise from previous error as follows:

```
except json.JSONDecodeError as e:
     raise ValueError("Invalid JSON in include-metadata or exclude-metadata") from e
```

ErnestoLoma commented on connectors/anaplan/app/clients.py:
```diff
@@ -0,0 +1,98 @@
+
+        if self.auth_type == "basic":
+            if not self.username or not self.password:
+                raise Exception(
```
I think you should raise a specific exception like `application_sdk.common.error_codes.ClientError` not `Exception`

ErnestoLoma commented on connectors/anaplan/app/clients.py:
```diff
@@ -0,0 +1,98 @@
+            else:
+                error_msg = f"Authentication failed with status {response.status_code if response else 'No response'}: {response.text if response else 'No response'}"
+                logger.error(error_msg)
+                raise Exception(error_msg)
```
I think you should raise a specific exception like `application_sdk.common.error_codes.ClientError` not `Exception`

ErnestoLoma commented on connectors/anaplan/app/clients.py:
```diff
@@ -0,0 +1,98 @@
+                logger.error(error_msg)
+                raise Exception(error_msg)
+        else:
+            raise Exception(f"Unsupported authentication type: {self.auth_type}")
```
I think you should raise a specific exception like `application_sdk.common.error_codes.ClientError` not `Exception`

ErnestoLoma commented on connectors/anaplan/app/clients.py:
```diff
@@ -0,0 +1,98 @@
+                token_info = token_data.get("tokenInfo", {})
+                self.auth_token = token_info.get("tokenValue")
+                if not self.auth_token:
+                    raise Exception("Token not found in authentication response")
```
I think you should raise a specific exception like `application_sdk.common.error_codes.ClientError` not `Exception`

ErnestoLoma commented on connectors/anaplan/app/handlers.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        try:
+            if not self.client:
+                raise Exception("App client not initialized")
```
I think you should raise a specific exception like `application_sdk.common.error_codes.ClientError` not `Exception`

ErnestoLoma commented on connectors/anaplan/app/handlers.py:
```diff
@@ -0,0 +1,234 @@
+            logger.error(f"Failed to fetch metadata: {str(e)}")
+            # For errors, we should raise an exception rather than return a list
+            # The SDK will handle the exception and return appropriate error response
+            raise Exception(f"Failed to fetch metadata: {str(e)}")
```
I think you should raise a specific exception like `application_sdk.common.error_codes.ClientError` not `Exception`
